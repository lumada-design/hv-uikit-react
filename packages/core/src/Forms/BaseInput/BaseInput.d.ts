import { StandardProps, InputProps } from "@material-ui/core";

export interface HvBaseInputProps
  extends StandardProps<InputProps, HvBaseInputClassKey, "onChange">{
  /**
   * Called back when the value is changed.
   */
  onChange?: (
    event: React.ChangeEvent<HTMLTextAreaElement | HTMLInputElement> | undefined,
    value: string
  ) => string;
  /**
   * If true, the input element will be required.
   */
  required?: boolean;
  /**
   * Denotes if the input is in an invalid state.
   */
  invalid?: boolean;
}

export type HvBaseInputClassKey =
  | "root"
  | "inputRoot"
  | "inputRootDisabled"
  | "inputRootInvalid"
  | "inputRootFocused"
  | "input"
  | "inputDisabled"
  | "multiLine"
  | "@global";

export default function HvBaseInput(props: HvBaseInputProps): JSX.Element | null;
