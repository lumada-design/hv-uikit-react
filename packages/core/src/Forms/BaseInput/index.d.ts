export { default } from "./BaseInput";
export * from "./BaseInput";
