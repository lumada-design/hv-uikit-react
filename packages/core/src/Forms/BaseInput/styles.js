import { outlineStyles } from "../../Focus/styles";

const styles = theme => ({
  root: {
    minWidth: "150px",
    maxWidth: "610px"
  },
  inputRoot: {
    margin: 0,
    width: "100%",
    borderStyle: "solid",
    borderWidth: " 0 0 1px 0",
    backgroundColor: theme.hv.palette.atmosphere.atmo1,
    borderColor: theme.hv.palette.atmosphere.atmo6,
    "&:hover": {
      borderStyle: "solid",
      borderWidth: " 0 0 1px 0",
      borderColor: theme.hv.palette.accent.acce1,
      background: theme.hv.palette.atmosphere.atmo1
    }
  },
  inputRootDisabled: {
    borderColor: theme.hv.palette.atmosphere.atmo6,
    background: theme.hv.palette.atmosphere.atmo4,
    "&:hover": {
      borderColor: theme.hv.palette.atmosphere.atmo6,
      background: theme.hv.palette.atmosphere.atmo4,
      cursor: "not-allowed"
    },
    cursor: "not-allowed"
  },
  inputRootFocused: {
    borderStyle: "solid",
    borderWidth: " 0 0 1px 0",
    borderColor: theme.hv.palette.accent.acce1,
    background: theme.hv.palette.atmosphere.atmo1,
    ...outlineStyles,
    "&:hover": {
      borderStyle: "solid",
      borderWidth: "0 0 1px 0",
      borderColor: theme.hv.palette.accent.acce1,
      background: theme.hv.palette.atmosphere.atmo1
    }
  },
  inputRootInvalid: {
    borderColor: theme.hv.palette.semantic.sema4,
    "&:hover": {
      borderColor: theme.hv.palette.semantic.sema4
    }
  },
  input: {
    height: `20px`,
    marginLeft: theme.hv.spacing.xs,
    marginRight: theme.hv.spacing.xs,
    padding: "6px 0 5px",
    ...theme.hv.typography.normalText,
    overflow: "hidden",
    textOverflow: "ellipsis",
    "&::placeholder": {
      ...theme.hv.typography.placeholderText,
      opacity: 1
    },
    "&::-ms-clear": {
      display: "none"
    }
  },
  inputDisabled: {
    cursor: "not-allowed"
  },
  multiLine: {
    padding: 0
  },
  "@global": {
    "input:-webkit-autofill": {
      "-webkit-box-shadow": `0 0 0px 1000px ${theme.hv.palette.atmosphere.atmo1} inset`,
      "-webkit-text-fill-color": theme.hv.typography.normalText.color
    }
  }
});

export default styles;
