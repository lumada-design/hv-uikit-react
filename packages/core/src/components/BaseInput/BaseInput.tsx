import { RefObject, useContext } from "react";
import {
  InputBaseComponentProps as MuiInputBaseComponentProps,
  InputProps as MuiInputProps,
} from "@mui/material";
import { HvBaseProps } from "../../types";
import {
  StyledRoot,
  StyledInputBorderContainer,
  StyledInput,
} from "./BaseInput.styles";
import {
  HvFormElementContext,
  HvFormElementDescriptorsContext,
  buildFormElementPropsFromContext,
  buildAriaPropsFromContext,
} from "../Forms/FormElement";
import clsx from "clsx";

export type HvBaseInputProps = Omit<MuiInputProps, "onChange"> &
  HvBaseProps<HTMLDivElement, { onChange }> & {
    /** The input name. */
    name?: string;
    /** The value of the input, when controlled. */
    value?: string;
    /** The initial value of the input, when uncontrolled. */
    defaultValue?: string;
    /** If `true` the input is disabled. */
    disabled?: boolean;
    /** Indicates that the input is not editable. */
    readOnly?: boolean;
    /** If true, the input element will be required. */
    required?: boolean;
    /** The function that will be executed onChange, allows modification of the input,
     * it receives the value. If a new value should be presented it must returned it. */
    onChange?: (
      event: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>,
      value: string
    ) => void;
    /** The input type. */
    type?: string;
    /** Label inside the input used to help user. */
    placeholder?: string;
    /** If true, a textarea element will be rendered. */
    multiline?: boolean;
    /** If true and multiline is also true the textarea element will be resizable. */
    resizable?: boolean;
    /** Denotes if the input is in an invalid state. */
    invalid?: boolean;
    /** Attributes applied to the input element. */
    inputProps?: MuiInputBaseComponentProps;
    /** Allows passing a ref to the underlying input */
    inputRef?: RefObject<HTMLElement>;
    /** A Jss Object used to override or extend the styles applied to the empty state component. */
    classes?: {
      /** Styles applied to the root container of the input. */
      root?: string;
      /** Styles applied to the root container of the input when it is disabled. */
      disabled?: string;
      /** Styles applied to the root container of the input when it is invalid. */
      invalid?: string;
      /** Styles applied to the root container of the input when it is resizable. */
      resizable?: string;
      /** Styles applied to input root which is the input that encloses all the other elements. */
      inputRoot?: string;
      /** Styles applied to input root when it is focused. */
      inputRootFocused?: string;
      /** Styles applied to input html element when it is disabled. */
      inputRootDisabled?: string;
      /** Styles applied to input html element when it is multiline mode. */
      inputRootMultiline?: string;
      /** Styles applied to input html element. */
      input?: string;
      /** Styles applied to input html element when is disabled. */
      inputDisabled?: string;
      /** Styles applied to input html element when it is resizable. */
      inputResizable?: string;
      /** Styles applied to the container of the border element. */
      inputBorderContainer?: string;
      /** Styles applied to the container of the border element, when in read only mode. */
      readOnly?: string;
    };
  };

/**
 * An Input component that only posses the most basic functionalities.
 * It should be used alongside the other form elements to construct a proper accessible form.
 */
export const HvBaseInput = ({
  classes,
  className = "",
  id,
  name,
  value,
  defaultValue,
  required,
  readOnly,
  disabled,
  onChange,
  type = "text",
  placeholder,
  multiline = false,
  resizable = false,
  invalid = false,
  inputRef,
  inputProps = {},
  ...others
}: HvBaseInputProps) => {
  const formElementContext = useContext(HvFormElementContext);
  const formElementProps = buildFormElementPropsFromContext(
    name,
    disabled,
    readOnly,
    required,
    status,
    formElementContext
  );

  const localInvalid = invalid || formElementProps.status === "invalid";

  const formElementDescriptorsContext = useContext(
    HvFormElementDescriptorsContext
  );
  const ariaProps = buildAriaPropsFromContext(
    inputProps,
    formElementDescriptorsContext,
    localInvalid,
    id
  );

  const onChangeHandler = (
    event: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
  ) => {
    onChange?.(event, event.target.value);
  };

  return (
    <StyledRoot
      className={clsx(
        "root",
        classes?.root,
        className,
        formElementProps.disabled && classes?.disabled,
        localInvalid && classes?.invalid,
        multiline && resizable && classes?.resizable,
        readOnly && classes?.readOnly
      )}
      $disabled={formElementProps.disabled}
      $invalid={localInvalid}
      $resizable={multiline && resizable}
      $readOnly={!!readOnly}
    >
      <StyledInput
        id={id}
        name={formElementProps.name}
        value={value}
        defaultValue={defaultValue}
        type={type}
        placeholder={placeholder}
        readOnly={!!formElementProps.readOnly}
        disabled={formElementProps.disabled}
        onChange={onChangeHandler}
        className={clsx(
          localInvalid && "inputRootInvalid",
          readOnly && "inputRootReadOnly"
        )}
        classes={{
          root: clsx("inputRoot", classes?.inputRoot),
          focused: clsx("inputRootFocused", classes?.inputRootFocused),
          disabled: clsx("inputRootDisabled", classes?.inputRootDisabled),
          multiline: clsx("inputRootMultiline", classes?.inputRootMultiline),
          input: clsx(
            "input",
            classes?.input,
            !formElementProps.disabled &&
              resizable &&
              clsx("inputResizable", classes?.inputResizable),
            disabled && clsx("inputDisabled", classes?.inputDisabled),
            readOnly && "readOnly" && classes?.readOnly
          ),
        }}
        inputProps={{
          // Avoid the required attribute at the root node
          required: formElementProps.required,
          ...inputProps,
          ...ariaProps,
        }}
        inputRef={inputRef}
        multiline={multiline}
        rows={10}
        $resizable={!formElementProps.disabled && resizable}
        $disabled={!!disabled}
        $readOnly={!!readOnly}
        $invalid={localInvalid}
        {...others}
      />
      {!multiline && (
        <StyledInputBorderContainer
          role="presentation"
          className={clsx(
            "inputBorderContainer",
            classes?.inputBorderContainer
          )}
        />
      )}
    </StyledRoot>
  );
};
