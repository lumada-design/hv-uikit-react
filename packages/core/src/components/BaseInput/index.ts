export { default as baseInputClasses } from "./baseInputClasses";
export * from "./baseInputClasses";
export * from "./BaseInput";
