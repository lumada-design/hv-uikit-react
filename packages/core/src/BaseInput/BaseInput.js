import React, { useContext } from "react";
import PropTypes from "prop-types";
import clsx from "clsx";
import { Input, withStyles } from "@material-ui/core";
import { refType } from "../utils";
import styles from "./styles";

import {
  HvFormElementContext,
  HvFormElementDescriptorsContext,
  buildFormElementPropsFromContext,
  buildAriaPropsFromContext,
} from "../Forms/FormElement";

/**
 * An Input component that only posses the most basic functionalities.
 * It should be used alongside the other form elements to construct a proper accessible form.
 */
const HvBaseInput = (props) => {
  const {
    classes,
    className = "",

    id,
    name,

    value,
    defaultValue,

    required,
    readOnly,
    disabled,

    onChange,

    type = "text",

    placeholder,

    multiline = false,
    resizable = false,
    invalid = false,

    inputRef,
    inputProps = {},

    ...others
  } = props;

  const formElementContext = useContext(HvFormElementContext);
  const formElementProps = buildFormElementPropsFromContext(props, formElementContext);

  const localInvalid = invalid || formElementProps.status === "invalid";

  const formElementDescriptorsContext = useContext(HvFormElementDescriptorsContext);
  const ariaProps = buildAriaPropsFromContext(
    inputProps,
    formElementDescriptorsContext,
    localInvalid,
    id
  );

  const onChangeHandler = (event) => {
    onChange?.(event, event.target.value);
  };

  return (
    <div
      className={clsx(classes.root, className, {
        [classes.disabled]: formElementProps.disabled,
        [classes.invalid]: localInvalid,
        [classes.resizable]: multiline && resizable,
        [classes.readOnly]: readOnly,
      })}
    >
      <Input
        id={id}
        name={formElementProps.name}
        value={value}
        defaultValue={defaultValue}
        type={type}
        placeholder={placeholder}
        readOnly={formElementProps.readOnly}
        disabled={formElementProps.disabled}
        onChange={onChangeHandler}
        classes={{
          root: classes.inputRoot,
          focused: classes.inputRootFocused,
          disabled: classes.inputRootDisabled,
          multiline: classes.inputRootMultiline,
          input: clsx(classes.input, {
            [classes.inputResizable]: !formElementProps.disabled && resizable,
            [classes.inputDisabled]: disabled,
          }),
        }}
        inputProps={{
          // avoid the required attribute at the root node
          required: formElementProps.required,
          ...inputProps,
          ...ariaProps,
        }}
        inputRef={inputRef}
        multiline={multiline}
        rows={10}
        {...others}
      />
      {!multiline && <div role="presentation" className={classes.inputBorderContainer} />}
    </div>
  );
};

HvBaseInput.formElementType = "input";

HvBaseInput.propTypes = {
  /**
   * Class names to be applied.
   */
  className: PropTypes.string,
  /**
   * A Jss Object used to override or extend the component styles applied.
   */
  classes: PropTypes.shape({
    /**
     * Styles applied to the root container of the input.
     */
    root: PropTypes.string,
    /**
     * Styles applied to the root container of the input when it is disabled.
     */
    disabled: PropTypes.string,
    /**
     * Styles applied to the root container of the input when it is invalid.
     */
    invalid: PropTypes.string,
    /**
     * Styles applied to the root container of the input when it is resizable.
     */
    resizable: PropTypes.string,

    /**
     * Styles applied to input root which is the input that encloses all the other elements.
     */
    inputRoot: PropTypes.string,
    /**
     * Styles applied to input root when it is focused.
     */
    inputRootFocused: PropTypes.string,
    /**
     * Styles applied to input html element when it is disabled.
     */
    inputRootDisabled: PropTypes.string,
    /**
     * Styles applied to input html element when it is multiline mode.
     */
    inputRootMultiline: PropTypes.string,

    /**
     * Styles applied to input html element.
     */
    input: PropTypes.string,
    /**
     * Styles applied to input html element when is disabled.
     */
    inputDisabled: PropTypes.string,
    /**
     * Styles applied to input html element when it is resizable.
     */
    inputResizable: PropTypes.string,

    /**
     * Styles applied to the container of the border element.
     */
    inputBorderContainer: PropTypes.string,
    /**
     * Styles applied to the container of the border element, when in read only mode.
     */
    readOnly: PropTypes.string,
  }).isRequired,

  /**
   * Id to be applied to the root node.
   */
  id: PropTypes.string,
  /**
   * The input name.
   */
  name: PropTypes.string,

  /**
   * The value of the input, when controlled.
   */
  value: PropTypes.string,
  /**
   * The initial value of the input, when uncontrolled.
   */
  defaultValue: PropTypes.string,

  /**
   * If `true` the input is disabled.
   */
  disabled: PropTypes.bool,
  /**
   * Indicates that the input is not editable.
   */
  readOnly: PropTypes.bool,
  /**
   * If true, the input element will be required.
   */
  required: PropTypes.bool,

  /**
   * The function that will be executed onChange, allows modification of the input,
   * it receives the value. If a new value should be presented it must returned it.
   */
  onChange: PropTypes.func,

  /**
   * The input type.
   */
  type: PropTypes.string,

  /**
   * Label inside the input used to help user.
   */
  placeholder: PropTypes.string,

  /**
   * If true, a textarea element will be rendered.
   */
  multiline: PropTypes.bool,
  /**
   * If true and multiline is also true the textarea element will be resizable.
   */
  resizable: PropTypes.bool,
  /**
   * Denotes if the input is in an invalid state.
   */
  invalid: PropTypes.bool,

  /**
   * Attributes applied to the input element.
   */
  inputProps: PropTypes.instanceOf(Object),
  /**
   * Allows passing a ref to the underlying input
   */
  inputRef: refType,
};

export default withStyles(styles, { name: "HvBaseInput" })(HvBaseInput);
