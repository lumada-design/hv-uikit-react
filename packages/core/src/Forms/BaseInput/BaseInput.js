import React, { useContext } from "react";
import PropTypes from "prop-types";
import clsx from "clsx";
import { Input, withStyles } from "@material-ui/core";
import { setId } from "../../utils";
import { HvFormElementContext } from "../FormElement";
import styles from "./styles";

const concatDescribeBy = descriptors => {
  let result;
  descriptors.forEach(descriptor => {
    if (descriptor && !result) {
      result = descriptor;
      return;
    }
    if (descriptor) result = result.concat(` ${descriptor}`);
  });
  return result;
};

/**
 * An Input component that only posses the most basic functionalities.
 * It should be used alongside the other form elements to construct a proper accessible form.
 */
const HvBaseInput = props => {
  const {
    placeholder,
    classes,
    className = "",
    id,
    disabled = false,
    inputProps,
    inputRef,
    onChange,
    value,
    multiline = false,
    defaultValue,
    required = false,
    invalid = false,
    ...others
  } = props;
  const {
    elementId,
    elementName,
    elementStatus,
    elementValue,
    elementDisabled,
    descriptors = {}
  } = useContext(HvFormElementContext);

  const onChangeHandler = event => {
    onChange?.(event, event.target.value);
  };

  const localInvalid = invalid || elementStatus === "invalid";
  const localValue = value ?? elementValue;
  const localDisabled = disabled || elementDisabled;
  const localId = id ?? setId(elementId, "input");
  const { HvLabel, HvSuggestions, HvInfoMessage, HvWarningText } = descriptors;

  return (
    <div
      className={clsx(classes.root, {
        [classes.disabledRoot]: localDisabled,
        [classes.invalidRoot]: localInvalid
      })}
    >
      <Input
        id={localId}
        name={elementName}
        defaultValue={defaultValue}
        value={localValue}
        disabled={localDisabled}
        placeholder={placeholder}
        onChange={onChangeHandler}
        classes={{
          input: classes.input,
          focused: classes.inputRootFocused,
          disabled: classes.inputDisabled,
          multiline: classes.multiLine
        }}
        className={clsx(classes.inputRoot, className, {
          [classes.inputRootDisabled]: localDisabled
        })}
        inputProps={{
          required,
          "aria-required": required || undefined,
          "aria-invalid": localInvalid || undefined,
          "aria-describedby": concatDescribeBy([HvWarningText?.[0]?.id, HvInfoMessage?.[0]?.id]),
          "aria-labelledby": HvLabel?.[0]?.id,
          "aria-controls": HvSuggestions?.[0]?.id,
          ...inputProps
        }}
        inputRef={inputRef}
        multiline={multiline}
        {...others}
      />
      <div role="presentation" className={classes.inputBorderContainer} />
    </div>
  );
};

HvBaseInput.propTypes = {
  /**
   * Class names to be applied.
   */
  className: PropTypes.string,
  /**
   * Id to be applied to the root node.
   */
  id: PropTypes.string,
  /**
   * A Jss Object used to override or extend the component styles applied.
   */
  classes: PropTypes.shape({
    /**
     * Styles applied to the root container of the input.
     */
    root: PropTypes.string,
    /**
     * Styles applied to the root container of the input when it is disabled.
     */
    disabledRoot: PropTypes.string,
    /**
     * Styles applied to the root container of the input when it is invalid.
     */
    invalidRoot: PropTypes.string,
    /**
     * Styles applied to input root which is the input that encloses all the other elements.
     */
    inputRoot: PropTypes.string,
    /**
     * Styles applied to the container of the border element.
     */
    inputBorderContainer: PropTypes.string,
    /**
     * Styles applied to the border element.
     */
    inputLowerBorder: PropTypes.string,
    /**
     * Styles applied to input root when it is disabled.
     */
    inputRootDisabled: PropTypes.string,
    /**
     * Styles applied to input root when it is focused.
     */
    inputRootFocused: PropTypes.string,
    /**
     * Styles applied to input html element.
     */
    input: PropTypes.string,
    /**
     * Styles applied to input html element when it is disabled.
     */
    inputDisabled: PropTypes.string,
    /**
     * Styles applied to input html element when it is multiline mode.
     */
    multiLine: PropTypes.string,
    /**
     * IE11 specific styling.
     */
    "@global": PropTypes.string
  }).isRequired,
  /**
   * Label inside the input used to help user.
   */
  placeholder: PropTypes.string,
  /**
   * Attributes applied to the input element.
   */
  inputProps: PropTypes.instanceOf(Object),
  /**
   * Allows passing a ref to the underlying input
   */
  inputRef: PropTypes.shape({ current: PropTypes.any }),
  /**
   * If ´true´ the input is disabled.
   */
  disabled: PropTypes.bool,
  /**
   * The function that will be executed onChange, allows modification of the input,
   * it receives the value. If a new value should be presented it must returned it.
   */
  onChange: PropTypes.func,
  /**
   * The value of the input, when controlled.
   */
  value: PropTypes.string,
  /**
   * The initial value of the input, when uncontrolled.
   */
  defaultValue: PropTypes.string,
  /**
   * If true, a textarea element will be rendered.
   */
  multiline: PropTypes.bool,
  /**
   * If true, the input element will be required.
   */
  required: PropTypes.bool,
  /**
   * Denotes if the input is in an invalid state.
   */
  invalid: PropTypes.bool
};

export default withStyles(styles, { name: "HvBaseInput" })(HvBaseInput);
