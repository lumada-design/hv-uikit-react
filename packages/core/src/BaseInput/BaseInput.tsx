import { forwardRef, useContext } from "react";
import MuiInputBase, { type InputBaseProps } from "@mui/material/InputBase";
import { useForkRef } from "@mui/material/utils";
import {
  useDefaultProps,
  type ExtractNames,
} from "@hitachivantara/uikit-react-utils";

import {
  buildAriaPropsFromContext,
  buildFormElementPropsFromContext,
  HvFormElementContext,
  HvFormElementDescriptorsContext,
} from "../FormElement";
import { staticClasses, useClasses } from "./BaseInput.styles";

export { staticClasses as baseInputClasses };

export type HvBaseInputClasses = ExtractNames<typeof useClasses>;

export interface HvBaseInputProps
  extends Omit<
    InputBaseProps,
    "onChange" | "classes" | "ref" | "color" | "size" | "inputProps"
  > {
  /** The input name. */
  name?: string;
  /** The value of the input, when controlled. */
  value?: React.InputHTMLAttributes<HTMLInputElement>["value"];
  /** The initial value of the input, when uncontrolled. */
  defaultValue?: React.InputHTMLAttributes<HTMLInputElement>["value"];
  /** If `true` the input is disabled. */
  disabled?: boolean;
  /** Indicates that the input is not editable. */
  readOnly?: boolean;
  /** If true, the input element will be required. */
  required?: boolean;
  /**
   * Callback fired when the value is changed.
   *
   * You can pull out the new value by accessing `event.target.value` (string),
   * or using the second callback argument.
   */
  onChange?: (
    event: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>,
    value: string,
  ) => void;
  /** If true, a textarea element will be rendered. */
  multiline?: boolean;
  /** If true and multiline is also true the textarea element will be resizable. */
  resizable?: boolean;
  /** Denotes if the input is in an invalid state. */
  invalid?: boolean;
  /**
   * Allows passing a ref to the underlying input
   * @deprecated Use `ref` directly instead
   * */
  inputRef?: InputBaseProps["inputRef"];
  /**
   * [Attributes](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#Attributes) applied to the `input` element.
   */
  inputProps?: React.InputHTMLAttributes<
    HTMLInputElement | HTMLTextAreaElement
  >;
  /** A Jss Object used to override or extend the styles applied to the component. */
  classes?: HvBaseInputClasses;
}

/**
 * An Input component that only posses the most basic functionalities.
 * It should be used alongside the other form elements to construct a proper accessible form.
 */
export const HvBaseInput = forwardRef<
  // no-indent
  React.ElementRef<"input">,
  HvBaseInputProps
>(function HvBaseInput(props, ref) {
  const {
    classes: classesProp,
    className = "",
    id,
    name,
    value,
    defaultValue,
    required,
    readOnly,
    disabled,
    onChange,
    type = "text",
    placeholder,
    multiline,
    resizable,
    invalid: invalidProp,
    inputRef,
    inputProps = {},
    ...others
  } = useDefaultProps("HvBaseInput", props);
  const { classes, cx } = useClasses(classesProp);
  const formElementContext = useContext(HvFormElementContext);
  const formElementProps = buildFormElementPropsFromContext(
    name,
    disabled,
    readOnly,
    required,
    formElementContext,
  );

  const forkedRef = useForkRef(ref, inputRef);

  const invalid = invalidProp || formElementProps.status === "invalid";

  const formElementDescriptorsContext = useContext(
    HvFormElementDescriptorsContext,
  );
  const ariaProps = buildAriaPropsFromContext(
    inputProps,
    formElementDescriptorsContext,
    invalid,
    id,
  );

  return (
    <MuiInputBase
      id={id}
      name={formElementProps.name}
      value={value}
      defaultValue={defaultValue}
      placeholder={placeholder}
      readOnly={!!formElementProps.readOnly}
      disabled={formElementProps.disabled}
      onChange={(event) => onChange?.(event, event.target.value)}
      className={cx(classes.root, className, {
        [classes.multiline]: multiline,
        [classes.invalid]: invalid,
        [classes.readOnly]: formElementProps.readOnly,
        [classes.disabled]: formElementProps.disabled,
      })}
      classes={{
        focused: classes.focused,
        input: cx(classes.input, {
          [classes.inputResizable]: !formElementProps.disabled && resizable,
          [classes.inputDisabled]: formElementProps.disabled,
          [classes.inputReadOnly]: formElementProps.readOnly,
        }),
      }}
      inputProps={{
        // Avoid the required attribute at the root node
        required: formElementProps.required,
        ...inputProps,
        ...ariaProps,
      }}
      inputRef={forkedRef}
      multiline={multiline}
      {...(multiline ? { rows: 10 } : { type })}
      {...others}
    />
  );
});
