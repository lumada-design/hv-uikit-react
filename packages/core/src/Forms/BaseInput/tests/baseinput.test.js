/* eslint-env jest */

import React from "react";
import { mount } from "enzyme";

import HvProvider from "../../../Provider";
import BaseInput from "..";

const inputProps = {
  "aria-label": "Input Label",
};

describe("v3 Input", () => {
  let wrapper;

  const getInputProps = (ParentElement) => ParentElement.find(BaseInput).props();

  beforeEach(async () => {
    wrapper = mount(
      <HvProvider>
        <BaseInput placeholder="test" inputProps={inputProps} />
      </HvProvider>
    );
  });

  it("should be defined", () => {
    expect(wrapper).toBeDefined();
  });

  it("should render correctly", () => {
    expect(wrapper.find(BaseInput)).toMatchSnapshot();
  });

  it("should render the Input component", () => {
    const inputComponent = wrapper.find(BaseInput);
    expect(inputComponent.length).toBe(1);
  });

  it("should disable the Base Input component", () => {
    wrapper = mount(
      <HvProvider>
        <BaseInput placeholder="test" disabled />
      </HvProvider>
    );
    expect(getInputProps(wrapper).disabled).toBe(true);
  });

  it("should render the Input component with the multiline style", () => {
    const wrapperTextArea = mount(
      <HvProvider>
        <BaseInput placeholder="test" multiline />
      </HvProvider>
    );
    expect(wrapperTextArea.find(BaseInput)).toMatchSnapshot();
  });
});
