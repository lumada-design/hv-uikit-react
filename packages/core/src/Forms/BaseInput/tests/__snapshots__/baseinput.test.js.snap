// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`v3 Input should render correctly 1`] = `
<WithStyles(HvBaseInput)
  inputProps={
    Object {
      "aria-label": "Input Label",
    }
  }
  placeholder="test"
>
  <HvBaseInput
    classes={
      Object {
        "disabledRoot": "HvBaseInput-disabledRoot",
        "input": "HvBaseInput-input",
        "inputBorderContainer": "HvBaseInput-inputBorderContainer",
        "inputDisabled": "HvBaseInput-inputDisabled",
        "inputRoot": "HvBaseInput-inputRoot",
        "inputRootDisabled": "HvBaseInput-inputRootDisabled",
        "inputRootFocused": "HvBaseInput-inputRootFocused",
        "invalidRoot": "HvBaseInput-invalidRoot",
        "multiLine": "HvBaseInput-multiLine",
        "root": "HvBaseInput-root",
      }
    }
    inputProps={
      Object {
        "aria-label": "Input Label",
      }
    }
    placeholder="test"
  >
    <div
      className="HvBaseInput-root"
    >
      <WithStyles(ForwardRef(Input))
        className="HvBaseInput-inputRoot"
        classes={
          Object {
            "disabled": "HvBaseInput-inputDisabled",
            "focused": "HvBaseInput-inputRootFocused",
            "input": "HvBaseInput-input",
            "multiline": "HvBaseInput-multiLine",
          }
        }
        inputProps={
          Object {
            "aria-controls": undefined,
            "aria-describedby": undefined,
            "aria-invalid": undefined,
            "aria-label": "Input Label",
            "aria-labelledby": undefined,
            "aria-required": undefined,
            "required": false,
          }
        }
        multiline={false}
        onChange={[Function]}
        placeholder="test"
      >
        <ForwardRef(Input)
          className="HvBaseInput-inputRoot"
          classes={
            Object {
              "colorSecondary": "MuiInput-colorSecondary",
              "disabled": "Mui-disabled HvBaseInput-inputDisabled",
              "error": "Mui-error",
              "focused": "Mui-focused HvBaseInput-inputRootFocused",
              "formControl": "MuiInput-formControl",
              "fullWidth": "MuiInput-fullWidth",
              "input": "MuiInput-input HvBaseInput-input",
              "inputMarginDense": "MuiInput-inputMarginDense",
              "inputMultiline": "MuiInput-inputMultiline",
              "inputTypeSearch": "MuiInput-inputTypeSearch",
              "marginDense": "MuiInput-marginDense",
              "multiline": "MuiInput-multiline HvBaseInput-multiLine",
              "root": "MuiInput-root",
              "underline": "MuiInput-underline",
            }
          }
          disableUnderline={true}
          inputProps={
            Object {
              "aria-controls": undefined,
              "aria-describedby": undefined,
              "aria-invalid": undefined,
              "aria-label": "Input Label",
              "aria-labelledby": undefined,
              "aria-required": undefined,
              "required": false,
            }
          }
          multiline={false}
          onChange={[Function]}
          placeholder="test"
        >
          <WithStyles(ForwardRef(InputBase))
            className="HvBaseInput-inputRoot"
            classes={
              Object {
                "colorSecondary": "MuiInput-colorSecondary",
                "disabled": "Mui-disabled HvBaseInput-inputDisabled",
                "error": "Mui-error",
                "focused": "Mui-focused HvBaseInput-inputRootFocused",
                "formControl": "MuiInput-formControl",
                "fullWidth": "MuiInput-fullWidth",
                "input": "MuiInput-input HvBaseInput-input",
                "inputMarginDense": "MuiInput-inputMarginDense",
                "inputMultiline": "MuiInput-inputMultiline",
                "inputTypeSearch": "MuiInput-inputTypeSearch",
                "marginDense": "MuiInput-marginDense",
                "multiline": "MuiInput-multiline HvBaseInput-multiLine",
                "root": "MuiInput-root",
                "underline": null,
              }
            }
            fullWidth={false}
            inputComponent="input"
            inputProps={
              Object {
                "aria-controls": undefined,
                "aria-describedby": undefined,
                "aria-invalid": undefined,
                "aria-label": "Input Label",
                "aria-labelledby": undefined,
                "aria-required": undefined,
                "required": false,
              }
            }
            multiline={false}
            onChange={[Function]}
            placeholder="test"
            type="text"
          >
            <ForwardRef(InputBase)
              className="HvBaseInput-inputRoot"
              classes={
                Object {
                  "adornedEnd": "MuiInputBase-adornedEnd",
                  "adornedStart": "MuiInputBase-adornedStart",
                  "colorSecondary": "MuiInputBase-colorSecondary MuiInput-colorSecondary",
                  "disabled": "Mui-disabled Mui-disabled HvBaseInput-inputDisabled",
                  "error": "Mui-error Mui-error",
                  "focused": "Mui-focused Mui-focused HvBaseInput-inputRootFocused",
                  "formControl": "MuiInputBase-formControl MuiInput-formControl",
                  "fullWidth": "MuiInputBase-fullWidth MuiInput-fullWidth",
                  "input": "MuiInputBase-input MuiInput-input HvBaseInput-input",
                  "inputAdornedEnd": "MuiInputBase-inputAdornedEnd",
                  "inputAdornedStart": "MuiInputBase-inputAdornedStart",
                  "inputHiddenLabel": "MuiInputBase-inputHiddenLabel",
                  "inputMarginDense": "MuiInputBase-inputMarginDense MuiInput-inputMarginDense",
                  "inputMultiline": "MuiInputBase-inputMultiline MuiInput-inputMultiline",
                  "inputTypeSearch": "MuiInputBase-inputTypeSearch MuiInput-inputTypeSearch",
                  "marginDense": "MuiInputBase-marginDense MuiInput-marginDense",
                  "multiline": "MuiInputBase-multiline MuiInput-multiline HvBaseInput-multiLine",
                  "root": "MuiInputBase-root MuiInput-root",
                }
              }
              fullWidth={false}
              inputComponent="input"
              inputProps={
                Object {
                  "aria-controls": undefined,
                  "aria-describedby": undefined,
                  "aria-invalid": undefined,
                  "aria-label": "Input Label",
                  "aria-labelledby": undefined,
                  "aria-required": undefined,
                  "required": false,
                }
              }
              multiline={false}
              onChange={[Function]}
              placeholder="test"
              type="text"
            >
              <div
                className="MuiInputBase-root MuiInput-root HvBaseInput-inputRoot"
                onClick={[Function]}
              >
                <input
                  aria-label="Input Label"
                  className="MuiInputBase-input MuiInput-input HvBaseInput-input"
                  onAnimationStart={[Function]}
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  placeholder="test"
                  required={false}
                  type="text"
                />
              </div>
            </ForwardRef(InputBase)>
          </WithStyles(ForwardRef(InputBase))>
        </ForwardRef(Input)>
      </WithStyles(ForwardRef(Input))>
      <div
        className="HvBaseInput-inputBorderContainer"
        role="presentation"
      />
    </div>
  </HvBaseInput>
</WithStyles(HvBaseInput)>
`;
